CORE-240: Re-Propagate updated public key

leviathanch opened the pull request:
The most significant change in this patch is the introduction of a new flag, called PEP_encrypt_flag_key_reset_repropagate which is meant to signify the engine, that a previously revoked key might be going to be needed in order to notify partners of the updated identity.

This however will obviously not work, when the other side has already received a revocation certificate and doesn't accept your key anymore, which is a case which will never happen, because if they've revoked your key, they certainly won't encrypt and email using it.

Commit:
CORE-240: Re-Propagate updated public key

The most significant change in this patch is the introduction of a
new flag, called PEP_encrypt_flag_key_reset_repropagate which is meant
to signify the engine, that a previously revoked key might be going to
be needed in order to notify partners of the updated identity.

This however will obviously not work, when the other side has already
received a revocation certificate and doesn't accept your key anymore,
which is a case which will never happen, because if they've revoked your
key, they certainly won't encrypt and email using it.

Commit:
Remove old keys

ig-garcia commented on src/message_api.h:
```diff
@@ -49,7 +49,11 @@ typedef enum _PEP_encrypt_flags {
+    PEP_encrypt_reencrypt = 0x40,
+
+    // Repropagation
+    PEP_encrypt_flag_key_reset_repropagate = 0x80
```
Would be nice if we can add some more info here in the comment.

leviathanch replied:
Runs through on my side...

uenceTest.o src/SigningTest.o src/SignOnlyTest.o src/SimpleBodyNotAltTest.o src/StickyBitTest.o src/StringlistTest.o src/StringpairListTest.o src/StrnstrTest.o src/SubkeyRatingEvalTest.o src/SyncWhenReceiveEncrypted.o src/TestSetDefaultFPRTest.o src/TestUtilities.o src/TrustManipulationTest.o src/TrustwordsTest.o src/UnencryptedPepMailTest.o src/UpdateIdAndMyselfTest.o src/URIAddressTest.o src/UserIDAliasTest.o src/UserIdCollisionTest.o src/VerifyTest.o -luuid -lpEpEngine -letpan -lstdc++ -lasn1 -luuid -lpthread -lpep_engine_sequoia_backend -lgtest -lgtest_main -lz
make[2]: Leaving directory '/home/leviathan/Desktop/Planck/sources/planckCoreV3/test'
rm -f -rf ./pEp_test_home/*
LD_LIBRARY_PATH=".:../src:$LD_LIBRARY_PATH" \
            PEP_NOABORT=noabort PEP_LOG=yes PEP_LOG_ASYNC=yes GTEST_COLOR=no \
            faketime '2021-07-07 05:05:30' \
             \
            python3 /home/leviathan/Desktop/Planck/sources/gtest-parallel/gtest-parallel \
        --gtest_color=no ./EngineTests
[587/587] StrnstrTest.check_strnstr_equal_but_size_too_small (8 ms)                                                                                                                                                                                                           
make[1]: Leaving directory '/home/leviathan/Desktop/Planck/sources/planckCoreV3/test'
leviathan@harvey:~/Desktop/Planck/sources/planckCoreV3>

ig-garcia commented on src/message_api.c:
```diff
@@ -2794,10 +2794,16 @@ static PEP_STATUS encrypt_message_possibly_with_media_key(
-    if (status != PEP_STATUS_OK)
-        goto pEp_error;
+
+    if ( flags & PEP_encrypt_flag_key_reset_repropagate ) {
```
So the purpose of the PEP_encrypt_flag_key_reset_repropagate flag is just making sure that myself is not called? So we will keep the old fpr?

ig-garcia replied:
I thought we were already using the old key for key reset messages? I think I missed something.

leviathanch replied:
Yes, you are missing that this is a repropagation, so we're signing with a revoked key.
This message has to be sent AFTER the key has been invalidated.

ip-plank-security replied:
exactly - the use case is to "silently" notify a peer who might have missed a key reset message that the peer is still using an old key.

ig-garcia replied:
Ok so we avoid the myself call so the revoked key is not replaced.

ghost commented:
Please make sure all tests pass when run together. If you introduce a newly created key, please be aware that the tests run with faketime in the past. So a present-day created key is not valid.

My complete error log is very long, but I can provide it if needed.

Summary:

```
DYLD_LIBRARY_PATH=".:../src:$DYLD_LIBRARY_PATH" \
            PEP_NOABORT=noabort PEP_LOG=yes PEP_LOG_ASYNC=yes GTEST_COLOR=no \
            faketime '2021-07-07 05:05:30' \
             \
            python3 /Users/dirk/projects/planck/json-runner/json-wrapper/gtest-parallel/gtest_parallel.py \
		--gtest_color=no ./EngineTests
[...]
[4/587] SyncWhenReceiveEncrypted.check_identity_after_reset returned with exit code 1 (299 ms)
[587/587] StringlistTest.check_stringlists (36 ms)
FAILED TESTS (4/587):
     302 ms: ./EngineTests SyncWhenReceiveEncrypted.create_test_mail
     297 ms: ./EngineTests SyncWhenReceiveEncrypted.decrypt_test_mail
     298 ms: ./EngineTests SyncWhenReceiveEncrypted.decrypt_test_mail_after_reset
     299 ms: ./EngineTests SyncWhenReceiveEncrypted.check_identity_after_reset
```

Sample error stack trace:

```
Error: No binding signature at time 2021-07-07T03:05:32Z: Invalid certificate
   0: backtrace::backtrace::libunwind::trace
             at /Users/dirk/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.67/src/backtrace/libunwind.rs:93:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/dirk/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.67/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at /Users/dirk/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.67/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/dirk/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.67/src/capture.rs:176:9
   3: backtrace::capture::Backtrace::new
             at /Users/dirk/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.67/src/capture.rs:140:22
   4: pep_engine_sequoia_backend::pgp_key_revoked::inner::{{closure}}
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreSequoiaBackend/src/pep.rs:33:34
   5: core::result::Result<T,E>::map_err
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/result.rs:829:27
   6: pep_engine_sequoia_backend::pgp_key_revoked::inner
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreSequoiaBackend/src/lib.rs:2178:14
   7: pep_engine_sequoia_backend::pgp_key_revoked::{{closure}}
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreSequoiaBackend/src/ffi.rs:40:23
   8: core::ops::function::FnOnce::call_once
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/ops/function.rs:250:5
   9: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panic/unwind_safe.rs:271:9
  10: std::panicking::try::do_call
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:504:40
  11: ___rust_try
  12: std::panicking::try
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:468:19
  13: std::panic::catch_unwind
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panic.rs:142:14
  14: pgp_key_revoked
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreSequoiaBackend/src/ffi.rs:39:21
  15: key_revoked
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreV3/src/pEpEngine.c:3507:11
  16: validate_fpr
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreV3/src/keymanagement.c:140:14
  17: set_own_key
             at /Users/dirk/projects/planck/json-runner/json-wrapper/planckCoreV3/src/keymanagement.c:2028:14
  18: _Z25set_up_ident_from_scratchP11_pEpSessionPKcS2_S2_S2_S2_PP13_pEp_identityb
             at src/TestUtilities.cc:132:18
  19: _ZN12_GLOBAL__N_124SyncWhenReceiveEncrypted9PrivAliceEv
             at src/SyncWhenReceiveEncrypted.cc:182:14
  20: _ZN56SyncWhenReceiveEncrypted_check_identity_after_reset_Test8TestBodyEv
             at src/SyncWhenReceiveEncrypted.cc:489:14
  21: __ZN7testing8internal38HandleSehExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc
  22: __ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc
  23: __ZN7testing4Test3RunEv
  24: __ZN7testing8TestInfo3RunEv
  25: __ZN7testing9TestSuite3RunEv
  26: __ZN7testing8internal12UnitTestImpl11RunAllTestsEv
  27: __ZN7testing8internal38HandleSehExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc
  28: __ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc
  29: __ZN7testing8UnitTest3RunEv
  30: __Z13RUN_ALL_TESTSv
  31: _main
```

ig-garcia commented on src/key_reset.c:
```diff
@@ -925,9 +931,15 @@ PEP_STATUS create_standalone_key_reset_message(PEP_SESSION session,
+       | PEP_encrypt_flag_force_no_attached_key;
+
+    if (!reset_msg->from->fpr) {
+        reset_msg->from->fpr = old_fpr;
```
Getting a warning from IDE: "Assigning to 'char *' from 'const char *' discards qualifiers". Probably nothing important.

ig-garcia commented on src/key_reset.c:
```diff
@@ -925,9 +931,15 @@ PEP_STATUS create_standalone_key_reset_message(PEP_SESSION session,
-                             &output_msg, PEP_enc_auto,
-                             PEP_encrypt_flag_key_reset_only | PEP_encrypt_flag_force_no_attached_key);
+    PEP_encrypt_flags_t flags =
+       PEP_encrypt_flag_key_reset_repropagate
```
I see we use the flag PEP_encrypt_flag_key_reset_repropagate every time on key reset, this means we will keep the old fpr in the "from" filed in all key reset messages (which I thought we were doing already).

ig-garcia replied:
I think I mistook the key reset itself with the creation of the key reset message.

planckdz reviewed (changes requested):
First round, will dig deeper.

planckdz commented on src/key_reset.c:
```diff
@@ -925,9 +931,15 @@ PEP_STATUS create_standalone_key_reset_message(PEP_SESSION session,
+       | PEP_encrypt_flag_force_no_attached_key;
+
+    if (!reset_msg->from->fpr) {
+        reset_msg->from->fpr = old_fpr;
```
`old_fpr` must be `strdup`ed, otherwise you risk double-frees.

leviathanch replied:
Ooops, that might be where the double free comes from. Good catch! Thanks!

planckdz commented on src/message_api.c:
```diff
@@ -6268,23 +6274,15 @@ static PEP_STATUS _decrypt_message(
-
-                                used_ids_for_key = identity_list_join(used_ids_for_key, used_cc_ids);
-                                
+                                used_ids_for_key = identity_list_add(used_ids_for_key, msg->from);
```
Why include `msg->from` here? The reset message later is hard-coded _to_ `msg->from`.

planckdz commented on test/src/KeyResetMessageTest.cc:
```diff
@@ -549,7 +549,7 @@ TEST_F(KeyResetMessageTest, check_receive_message_to_revoked_key_from_unknown) {
 }
 
-
+#if 0
```
Please remove tests that are completely obsolete, that is, they can't be adapted or it doesn't make sense. No `#if 0`.

planckdz commented on test/src/SyncWhenReceiveEncrypted.cc:
```diff
@@ -0,0 +1,503 @@
+// This file is under GNU General Public License 3.0
+// see LICENSE.txt
+
+#define UPDATE_DATA 0
```
Looks like a clever way to have the test create/update its own data. While I appreciate the idea, do you see a better way?

leviathanch replied:
It's the easiest way to update all the keys and emails being emitted, I don't really see a better way than that

leviathanch commented on src/key_reset.c:
```diff
@@ -925,9 +931,15 @@ PEP_STATUS create_standalone_key_reset_message(PEP_SESSION session,
+       | PEP_encrypt_flag_key_reset_only
+       | PEP_encrypt_flag_force_no_attached_key;
+
+    if (!reset_msg->from->fpr) {
```
Hmm. Good point

Commit:
CORE-240: Remove obsolete test

Commit:
CORE-240: Avoid double free

Commit:
CORE-240: Update test case

Add an assert condition, making sure the reset message actually
is being emitted.

Commit:
CORE-240: remove from

Adding the from identity isn't needed for getting it to work

Commit:
CORE-240: Add comment

Elaborate a bit more on PEP_encrypt_flag_key_reset_repropagate

ghost commented:
@leviathanch We need to agree on an exact faketime, otherwise the tests are prone to break, depending on OS and timezone.

This is the time the tests officially currently operate: https://github.com/plancksecurity/foundation-planckCoreV3/blob/develop/test/Makefile#L139

That is, unfortunately, local timezone, we should correct this, e.g. with `faketime '2021-07-07 00:00:00Z'`. But even then, at least one of your keys (test/test_keys/priv/swre_049C0ECD55C3E47F0A44BB43C8076095EA27D98E.asc) has a creation time of `2021-07-07 04:04:30 UTC`, which is in the future of this faketime.

So on my macOS all those tests fail like this:

```
Error: No binding signature at time 2021-07-07T00:00:01Z: Invalid certificate
```

Can we please do the following:

* Agree on an exact faketime parameter that covers all existing tests.
* Put this exact faketime paramater into the tests Makefile referenced above.
* Update your tests so that all keys have a creation date before that faketime (or, simply tweak the faketime paramater that all tests run).

ghost commented:
Some test files (incl. the source code) look like they have windows file endings. The core should stay at UNIX. Please fix.

(I do hope VS on windows understands this concept, in case one edits core source code in it).

ghost commented:
Faketime fix for macOS: https://github.com/plancksecurity/foundation-planckCoreV3/tree/dz/CORE-240-PR2

dirkz committed:
CORE-240 Fix faketime for macOS

dirkz committed:
CORE-240 Unix line endings

dirkz committed:
CORE-240 Remove trailing whitespace

saschabacardit commented on test/Makefile:
```diff
@@ -136,7 +136,7 @@ test: all prepare-test-data
 	$(TEST_CMD_PFX) \
             PEP_NOABORT=noabort PEP_LOG=yes PEP_LOG_ASYNC=yes GTEST_COLOR=no \
-            faketime '2021-07-07 05:05:30' \
+            faketime '2021-07-07 23:05:30Z' \
```
Why this change?

planckdz replied:
Because some of the (generated) test case keys have a generation date sometime on that day (021-07-07 ..., please see above for a fitting error message), and a relatively up-to-date macOS faketime needs a very precise time in order to work. This ensures that the keys used in some tests "have been generated in the past, and not in the future".

saschabacardit commented on test/src/KeyResetMessageTest.cc:
```diff
@@ -549,92 +549,6 @@ TEST_F(KeyResetMessageTest, check_receive_message_to_revoked_key_from_unknown) {
-    }
-}
-
-
```
Is this test replaced anywhere?

saschabacardit replied:
(I assume it is on the other test files and this was left as redundant mostly a confirmation)
